pkg/reconciler/tenancy/workspacedeletion/workspace_deletion_controller.go

```diff
@@ -246,35 +273,78 @@ func (c *Controller) patchCondition(ctx context.Context, old, new *tenancyv1alph
+			logger.Info("deleting cluster role bindings")
+			if err := c.kubeClusterClient.Cluster(clusterName).RbacV1().ClusterRoleBindings().DeleteCollection(ctx, backgroudDeletion, metav1.ListOptions{}); err != nil && !apierrors.IsNotFound(err) {
 				return fmt.Errorf("could not delete clusterrolebindings for workspace %s: %w", clusterName, err)
 			}
-			logger.V(2).Info("removing finalizer from ClusterWorkspace")
-			_, err := c.kcpClusterClient.TenancyV1alpha1().ClusterWorkspaces().Update(
-				logicalcluster.WithCluster(ctx, clusterName), workspace, metav1.UpdateOptions{})
+
+			if ws.Spec.Owner != nil {
+				gvr := schema.GroupVersionResource{
+					Resource: ws.Spec.Owner.Resource,
+				}
+				comps := strings.SplitN(ws.Spec.Owner.APIVersion, "/", 2)
+				if len(comps) == 2 {
+					gvr.Group = comps[0]
+					gvr.Version = comps[1]
+				} else {
+					gvr.Version = comps[0]
+				}
+				uid := types.UID(ws.Spec.Owner.UID)
+				logger = logger.WithValues("owner.gvr", gvr, "owner.uid", uid, "owner.name", ws.Spec.Owner.Name, "owner.namespace", ws.Spec.Owner.Namespace, "owner.cluster", ws.Spec.Owner.Cluster)
+
+				// remove finalizer from owner
+				logger.Info("checking owner for finalizer")
+				ownerCtx := logicalcluster.WithCluster(ctx, logicalcluster.New(ws.Spec.Owner.Cluster))
+				obj, err := c.dynamicFrontProxyClient.Resource(gvr).Namespace(ws.Spec.Owner.Namespace).Get(ownerCtx, ws.Spec.Owner.Name, metav1.GetOptions{})
+				if err != nil && !apierrors.IsNotFound(err) {
+					return fmt.Errorf("could not get owner %s %s/%s in cluster %s: %w", gvr, ws.Spec.Owner.Namespace, ws.Spec.Owner.Name, ws.Spec.Owner.Cluster, err)
+				} else if err == nil && obj.GetUID() != uid {
+					logger.Info("owner has changed, skipping finalizer removal")
+					return fmt.Errorf("could not get owner %s %s/%s in cluster %s is of wrong UID: %w", gvr, ws.Spec.Owner.Namespace, ws.Spec.Owner.Name, ws.Spec.Owner.Cluster, err)
+				} else if err == nil {
+					finalizers := sets.NewString(obj.GetFinalizers()...)
+					if finalizers.Has(tenancyv1alpha1.ThisWorkspaceFinalizer) {
+						logger.Info("removing finalizer from owner")
+						finalizers.Delete(tenancyv1alpha1.ThisWorkspaceFinalizer)
+						obj.SetFinalizers(finalizers.List())
+						if obj, err = c.dynamicFrontProxyClient.Resource(gvr).Namespace(ws.Spec.Owner.Namespace).Update(ownerCtx, obj, metav1.UpdateOptions{}); err != nil {
+							return fmt.Errorf("could not remove finalizer from owner %s %s/%s in cluster %s: %w", gvr, ws.Spec.Owner.Namespace, ws.Spec.Owner.Name, ws.Spec.Owner.Cluster, err)
+						}
+					}
+
+					// delete owner
+					if obj.GetDeletionTimestamp().IsZero() && ws.Spec.DirectlyDeletable {
+						logger.Info("deleting owner")
+						if err := c.dynamicFrontProxyClient.Resource(gvr).Namespace(ws.Spec.Owner.Namespace).Delete(ownerCtx, ws.Spec.Owner.Name, metav1.DeleteOptions{Preconditions: &metav1.Preconditions{UID: &uid}}); err != nil && !apierrors.IsNotFound(err) {
```

Comment:
interesting, isn't this the obj that triggers `ThisWorkspace` deletion? In that case, removing the Finalizer should be enough.

Reply:
This allows reverse deletion, i.e. the ThisWorkspace triggers the deletion of the owner.